internal/app/machined/pkg/controllers/block/disks.go

```diff
@@ -247,3 +261,62 @@ func (ctrl *DisksController) analyzeBlockDevice(
 		return nil
 	})
 }
+
+func serialFromUdevdHelpers(ctx context.Context, id, transport string) string {
+	switch strings.ToLower(transport) {
+	case "ata":
+		return runUdevdHelper(
+			ctx,
+			"/usr/lib/udev/ata_id",
+			"--export", filepath.Join("/dev", id),
+		)
+
+	case "scsi":
+		return runUdevdHelper(
+			ctx,
+			"/usr/lib/udev/scsi_id",
+			"--export", "--allowlisted", "--device", filepath.Join("/dev", id),
+		)
+
+	default:
+		return ""
+	}
+}
+
+func runUdevdHelper(ctx context.Context, helper string, args ...string) string {
+	out, err := cmd.RunContext(ctx, helper, args...)
+	if err != nil {
+		return ""
+	}
+
+	env := parseEnv(strings.NewReader(out))
+
+	return cmp.Or(
+		env["ID_SERIAL"],
+		env["ID_SERIAL_SHORT"],
+	)
+}
+
+func parseEnv(r io.Reader) map[string]string {
+	env := map[string]string{}
+
+	scanner := bufio.NewScanner(r)
+	for scanner.Scan() {
+		line := scanner.Text()
+		if line == "" || strings.HasPrefix(line, "#") {
+			continue
+		}
+
+		parts := strings.SplitN(line, "=", 2)
```

Comment:
nit, we can do more modern (and more allocation-friendly) `strings.Cut`

```suggestion
		key, value, ok := strings.Cut(line, "=")
		if !ok {
		   continue
		}
```